TST: byte_bounds contiguity handling

tylerjereddy opened the pull request:
Add unit tests to probe [uncovered](https://codecov.io/gh/numpy/numpy/src/master/numpy/lib/utils.py#L185) code paths in `byte_bounds()` that handle both positive and negative strides in non-contiguous arrays.

eric-wieser commented on numpy/lib/tests/test_utils.py:
```diff
@@ -56,11 +56,24 @@ def test_safe_eval_nameconstant():
+    def test_byte_bounds(self):
+        a = arange(12).reshape(3, 4)
+        low, high = utils.byte_bounds(a)
+        assert_equal(high - low, a.size * a.itemsize)
```
I think this could do with a comment explaining that this is because the array is contiguous

tylerjereddy replied:
added

eric-wieser commented on numpy/lib/tests/test_utils.py:
```diff
@@ -56,11 +56,24 @@ def test_safe_eval_nameconstant():
+        b = a.T[::-1]
+        low, high = utils.byte_bounds(b)
+        assert_equal(high - low, b.size * b.itemsize)
 
```
Needs an extra newline here

tylerjereddy replied:
added

eric-wieser commented on numpy/lib/tests/test_utils.py:
```diff
@@ -56,11 +56,24 @@ def test_safe_eval_nameconstant():
+        low, high = utils.byte_bounds(a)
+        assert_equal(high - low, a.size * a.itemsize)
+
+    def test_non_contiguous_positive_stride(self):
```
I'd argue that this array, and the one below, is contiguous, just in an unusual order. You should add a test for strided arrays with padding, perhaps using `[::2]`

tylerjereddy committed:
TST: test byte_bounds contiguity handling

* add unit tests to probe previously-uncovered
code paths in byte_bounds() that handle both
positive and negative strides in non-contiguous
arrays (or, at least, arrays with unusual ordering)

tylerjereddy commented:
Revised based on feedback. The contiguous terminology was from the source code, not I, but I've endeavoured to relax that a bit too.
